demos/selenium-test-app/lib.rs

```diff
@@ -0,0 +1,239 @@
+                    headers,
+                    body: Cow::Owned(ByteBuf::from(value.clone())),
+                }
+            })
+        }),
+        _ => {
+            headers.push(certificate_header);
+            ASSETS.with(|a| match a.borrow().get(path) {
+                Some((asset_headers, value)) => {
+                    headers.append(&mut asset_headers.clone());
+
+                    HttpResponse {
+                        status_code: 200,
+                        headers,
+                        body: Cow::Owned(ByteBuf::from(value.clone())),
+                    }
+                }
+                None => HttpResponse {
+                    status_code: 404,
+                    headers,
+                    body: Cow::Owned(ByteBuf::from(format!("Asset {} not found.", path))),
+                },
+            })
+        }
+    }
+}
+
+fn make_asset_certificate_header(
+    asset_hashes: &RbTree<&'static str, Hash>,
+    asset_name: &str,
+) -> (String, String) {
+    let certificate = api::data_certificate().unwrap_or_else(|| {
+        api::trap("data certificate is only available in query calls");
+    });
+    let witness = asset_hashes.witness(asset_name.as_bytes());
+    let tree = ic_certified_map::labeled(b"http_assets", witness);
+    let mut serializer = serde_cbor::ser::Serializer::new(vec![]);
+    serializer.self_describe().unwrap();
+    tree.serialize(&mut serializer)
+        .unwrap_or_else(|e| api::trap(&format!("failed to serialize a hash tree: {}", e)));
+    (
+        "IC-Certificate".to_string(),
+        format!(
+            "certificate=:{}:, tree=:{}:",
+            base64::encode(&certificate),
+            base64::encode(&serializer.into_inner())
+        ),
+    )
+}
+
+#[init]
+#[post_upgrade]
+pub fn init_assets() {
+    ASSETS.with(|a| {
+        ASSET_HASHES.with(|ah| {
+            let mut assets = a.borrow_mut();
+            let mut asset_hashes = ah.borrow_mut();
+            for (path, content, content_type) in get_assets() {
+                asset_hashes.insert(path, sha2::Sha256::digest(content).into());
+                let mut headers = vec![];
+                headers.push((
+                    "Content-Type".to_string(),
+                    content_type.to_mime_type_string(),
+                ));
+                assets.insert(path, (headers, content.to_vec()));
+            }
+            update_root_hash(&asset_hashes);
+        });
+    });
+}
+
+fn update_root_hash(a: &RbTree<&'static str, [u8; 32]>) {
+    let prefixed_root_hash = labeled_hash(b"http_assets", &a.root_hash());
+    api::set_certified_data(&prefixed_root_hash[..]);
+}
+
+lazy_static! {
+    // The full content of the index.html, after the canister ID (and script tag) have been
+    // injected
+    static ref INDEX_HTML_STR: String = {
+        let canister_id = api::id();
+        let index_html = include_str!("dist/index.html");
+        let index_html = index_html.replace(
```

Comment:
copy pasta?

Reply:
Yes, it has been copied from II but it is in use.

Reply:
So II's index.html has a `script` tag with id `setupJs` which gets looked up and replaced.

In this test app I can't see a script tag in the index.html. Is this using a `script` tag generated by webpack? Why does the new value have `id="setupJs"`?

Reply:
> Is this using a script tag generated by webpack?
Yes, exactly.

> Why does the new value have id="setupJs"?
This is indeed a copy paste artifact.